firmware/targets/f7/Src/update.c

```diff
@@ -0,0 +1,157 @@
+#include "update.h"
+
+#include <furi.h>
+#include <furi_hal.h>
+#include <flipper.h>
+
+#include <fatfs.h>
+#include <flipper_format/flipper_format.h>
+
+#include <updater/util/update_manifest.h>
+
+#define TAG "Main"
+
+static FATFS fs;
+
+static const char FS_ROOT_PATH[] = "/";
+static const char CONFIG_PATH[] = "/update/update.cfg";
+
+#define CHECK_FRESULT(result)   \
+    {                           \
+        if((result) != FR_OK) { \
+            return false;       \
+        }                       \
+    }
+
+static bool flipper_update_init() {
+    furi_hal_interrupt_init();
+    furi_hal_delay_init(false);
+
+    MX_GPIO_Init();
+
+    //furi_hal_bootloader_init();
+    //furi_hal_version_init();
+
+    furi_hal_spi_init();
+    furi_hal_crc_init();
+
+    MX_FATFS_Init();
+    if(!hal_sd_detect()) {
+        return false;
+    }
+
+    if(BSP_SD_Init(true)) {
+        return false;
+    }
+
+    CHECK_FRESULT(f_mount(&fs, FS_ROOT_PATH, 1));
+    return true;
+}
+
+static inline void flipper_update_target_switch(void* offset) {
+    asm volatile("ldr    r3, [%0]    \n"
+                 "msr    msp, r3     \n"
+                 "ldr    r3, [%1]    \n"
+                 "mov    pc, r3      \n"
+                 :
+                 : "r"(offset), "r"(offset + 0x4)
+                 : "r3");
+}
+
+static bool flipper_update_load_stage(UpdateManifest* manifest) {
+    FIL file;
+    FILINFO stat;
+
+    furi_hal_crc_reset();
+
+    string_t loader_img_path;
+    string_init_printf(
+        loader_img_path, "/update/%s", string_get_cstr(manifest->staged_loader_file));
+
+    CHECK_FRESULT(f_stat(string_get_cstr(loader_img_path), &stat));
```

Comment:
Leads to memory leak via loader_img_path

Reply:
Macro with return is dangerous.

Reply:
Fixed